python-avd/pyavd/get_fabric_documentation.py

```diff
@@ -102,3 +114,93 @@ def _get_p2p_links_csv(fabric_documentation_facts: FabricDocumentationFacts) ->
     )
     csv_content.seek(0)
     return csv_content.read()
+
+
+def _get_digital_twin(fabric_documentation_facts: FabricDocumentationFacts) -> ACTDigitalTwin | None:
+    digital_twin_env = next(
+        (
+            environment
+            for device_structurude_config in fabric_documentation_facts.structured_configs.values()
+            if (environment := get(device_structurude_config, "metadata.digital_twin.environment")) is not None
+        ),
+        None,
+    )
+    match digital_twin_env:
+        case "act":
+            return _get_digital_twin_act(fabric_documentation_facts)
+        case _:
+            return None
+
+
+def _get_digital_twin_act(fabric_documentation_facts: FabricDocumentationFacts) -> ACTDigitalTwin:
+    # Identify common username for fabric nodes
+    # Value is enforced as a non-empty string during the generation of the metadata part of the structured_config
+    digital_twin_fabric_username: str = next(
+        (
+            get(device_structured_config, "metadata.digital_twin.username")
+            for device_structured_config in fabric_documentation_facts.structured_configs.values()
+        ),
+    )
+
+    # Identify common password for fabric nodes
+    # Value is enforced as a non-empty string during the generation of the metadata part of the structured_config
+    digital_twin_fabric_password: str = next(
+        (
+            get(device_structured_config, "metadata.digital_twin.password")
+            for device_structured_config in fabric_documentation_facts.structured_configs.values()
+        ),
+    )
+
+    digital_twin_node_types: dict[str, ActNodeTypeSettings | None] = {
+        "cloudeos": None,
+        "cvp": None,
+        "generic": None,
+        "third-party": None,
+        "tools-server": None,
+        "veos": None,
+    }
+    digital_twin_devices: list[dict[str, ActNodeSettings]] = []
+    device_list: list[str] = list(fabric_documentation_facts.avd_facts)
+    for device in sorted(device_list):
+        if (
+            digital_twin_node_type := get(fabric_documentation_facts.structured_configs, f"{device}.metadata.digital_twin.node_type", "")
+        ) in digital_twin_node_types and not digital_twin_node_types[digital_twin_node_type]:
+            digital_twin_node_types[digital_twin_node_type] = ActNodeTypeSettings(username=digital_twin_fabric_username, password=digital_twin_fabric_password)
+
+        digital_twin_devices.append(
+            {
+                device: ActNodeSettings(
+                    # All three values are enforced as non-empty strings during the generation of the metadata part of the structured_config
+                    node_type=get(fabric_documentation_facts.structured_configs, f"{device}..metadata..digital_twin..node_type", separator=".."),
```

Comment:
do you need to get it again when you have set it line 166?
I would expect this to work:

```suggestion
                    node_type=digital_twin_node_type,
```

Reply:
True! Fixed now.